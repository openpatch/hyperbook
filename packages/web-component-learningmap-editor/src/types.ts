import { Node, Edge, Box } from "@xyflow/react";

// ============================================================================
// TYPES & INTERFACES
// ============================================================================

export interface UnlockCondition {
  after?: string[];
  date?: string;
  password?: string;
}

export interface CompletionNeed {
  id: string;
  source?: string;
  target?: string;
}

export interface Completion {
  needs?: CompletionNeed[];
  optional?: CompletionNeed[];
}

export interface NodeData {
  state: "locked" | "unlocked" | "started" | "completed" | "mastered";
  label: string;
  description?: string;
  duration?: string;
  unlock?: UnlockCondition;
  completion?: Completion;
  video?: string;
  resources?: { label: string; url: string }[];
  summary?: string;
  [key: string]: any;
}

export interface ImageNodeData {
  data?: string; // base64 encoded image
}

export interface TextNodeData {
  text?: string;
  fontSize?: number;
  color?: string;
  rotation?: number;
}

export type BackgroundNodeData = ImageNodeData | TextNodeData;

export interface BackgroundConfig {
  color?: string;
  nodes?: Node<BackgroundNodeData>[];
}

export interface EdgeConfig {
  animated?: boolean;
  color?: string;
  width?: number;
  type?: string;
}

export interface RoadmapData {
  nodes?: Node<NodeData>[];
  edges?: Edge[];
  background?: BackgroundConfig;
  edgeConfig?: EdgeConfig;
}

export type Orientation = "horizontal" | "vertical";

export type HelperLine = {
  // Used to filter out helper lines corresponding to the node being dragged
  node: Node;
  // We use it to check that the helper line is within the viewport.
  nodeBox: Box;
  // 0 for horizontal, 1 for vertical
  orientation: Orientation;
  // If orientation is 'horizontal', `position` holds the Y coordinate of the helper line.
  // (Might correspond to the top or bottom position of a node, or other anchors).
  // If orientation is 'vertical', `position` holds the X coordinate of the helper line.
  position: number;
  // Optional color for the helper line
  color?: string;
  anchorName: string;
};
