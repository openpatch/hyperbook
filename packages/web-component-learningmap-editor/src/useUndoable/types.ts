export type ActionType =
  | "undo"
  | "redo"
  | "update"
  | "reset"
  | "resetInitialState";

export type HistoryLimit = number | "infinium" | "infinity";

export type MutationBehavior =
  | "mergePastReversed"
  | "mergePast"
  | "destroyFuture"
  | "keepFuture";

export interface Action<T> {
  type: ActionType;
  payload?: T;
  behavior?: MutationBehavior;
  historyLimit?: HistoryLimit;
  ignoreIdenticalMutations?: boolean;
  cloneState?: boolean;
  ignoreAction?: boolean;
}

export interface State<T> {
  past: T[];
  present: T;
  future: T[];
}

export interface Options {
  behavior?: MutationBehavior;
  historyLimit?: HistoryLimit;
  ignoreIdenticalMutations?: boolean;
  cloneState?: boolean;
}

export type UseUndoable<T> = [
  T,
  (
    payload: T | ((oldValue: T) => T),
    behavior?: MutationBehavior,
    ignoreAction?: boolean,
  ) => void,
  {
    past: T[];
    future: T[];

    undo: () => void;
    canUndo: boolean;
    redo: () => void;
    canRedo: boolean;

    reset: (initialState?: T) => void;
    resetInitialState: (newInitialState: T) => void;
    static_setState: (
      payload: T,
      behavior?: MutationBehavior,
      ignoreAction?: boolean,
    ) => void;
  },
];
