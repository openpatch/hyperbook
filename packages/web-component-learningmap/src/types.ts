import { Node, Edge, Box } from "@xyflow/react";

// ============================================================================
// TYPES & INTERFACES
// ============================================================================

export interface UnlockCondition {
  after?: string[];
  date?: string;
  password?: string;
}

export interface Completion {
  needs?: string[];
  optional?: string[];
}

export interface NodeData {
  state: "locked" | "unlocked" | "started" | "completed" | "mastered";
  label: string;
  description?: string;
  duration?: string;
  unlock?: UnlockCondition;
  completion?: Completion;
  video?: string;
  resources?: { label: string; url: string }[];
  summary?: string;
  [key: string]: any;
}

export interface ImageNodeData {
  data?: string; // base64 encoded image
}

export interface TextNodeData {
  text?: string;
  fontSize?: number;
  color?: string;
  rotation?: number;
}

export type BackgroundNodeData = ImageNodeData | TextNodeData;

export interface BackgroundConfig {
  color?: string;
  nodes?: Node<BackgroundNodeData>[];
}

export interface Settings {
  title?: string;
  background?: BackgroundConfig;
  language?: string;
}

export interface EdgeConfig {
  animated?: boolean;
  color?: string;
  width?: number;
  type?: string;
}

export interface RoadmapData {
  nodes?: Node<NodeData>[];
  edges?: Edge[];
  settings: Settings;
  version: number;
}

export interface RoadmapState {
  nodes: Record<string, { state: string }>;
  x: number;
  y: number;
  zoom: number;
}

export type Orientation = "horizontal" | "vertical";

export type HelperLine = {
  // Used to filter out helper lines corresponding to the node being dragged
  node: Node;
  // We use it to check that the helper line is within the viewport.
  nodeBox: Box;
  // 0 for horizontal, 1 for vertical
  orientation: Orientation;
  // If orientation is 'horizontal', `position` holds the Y coordinate of the helper line.
  // (Might correspond to the top or bottom position of a node, or other anchors).
  // If orientation is 'vertical', `position` holds the X coordinate of the helper line.
  position: number;
  // Optional color for the helper line
  color?: string;
  anchorName: string;
};
