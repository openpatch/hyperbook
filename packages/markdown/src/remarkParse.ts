// @ts-nocheck
import { Root } from "mdast";
import { fromMarkdown } from "mdast-util-from-markdown";
import { Plugin } from "unified";
import micromarkSubSup from "./micromarkSubSup";
import mdastSubSup from "./mdastSubSup";

const remarkParse: Plugin<[], string, Root> = function () {
  const self = this;

  self.parser = function (doc) {
    const extensions = self.data("micromarkExtensions") || [];
    const mdastExtensions = self.data("fromMarkdownExtensions") || [];

    return fromMarkdown(doc, {
      ...self.data("settings"),
      extensions: [micromarkSubSup, ...extensions],
      mdastExtensions: [mdastSubSup, ...mdastExtensions],
    });
  };
};

export default remarkParse;
