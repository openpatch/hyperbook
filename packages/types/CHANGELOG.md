# @hyperbook/types

## 0.0.1

### Patch Changes

- [#316](https://github.com/openpatch/hyperbook/pull/316) [`cedb551`](https://github.com/openpatch/hyperbook/commit/cedb55191fd025b5a214df406a53cbab5d1b1bc1) Thanks [@mikebarkmin](https://github.com/mikebarkmin)! - Extract Hyperbook types into a separate package
