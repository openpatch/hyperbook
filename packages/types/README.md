# @hyperbook/types

## Installation

```sh
yarn add @hyperbook/types
# or
npm i @hyperbook/types
```
