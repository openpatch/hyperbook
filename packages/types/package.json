{
  "name": "@hyperbook/types",
  "version": "0.5.0",
  "author": "Mike Barkmin",
  "homepage": "https://github.com/openpatch/hyperbook#readme",
  "license": "MIT",
  "main": "dist/index.cjs.js",
  "module": "dist/index.esm.mjs",
  "types": "dist/index.d.ts",
  "typings": "dist/index.d.ts",
  "sideEffects": false,
  "exports": {
    ".": {
      "require": "./dist/index.cjs.js",
      "default": "./dist/index.esm.mjs"
    }
  },
  "files": [
    "dist"
  ],
  "publishConfig": {
    "access": "public"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/openpatch/hyperbook.git",
    "directory": "packages/types"
  },
  "bugs": {
    "url": "https://github.com/openpatch/hyperbook/issues"
  },
  "scripts": {
    "prebuild": "rimraf dist",
    "version": "pnpm build",
    "lint": "tsc --noEmit",
    "build": "pnpm build:pkg && pnpm build:types",
    "build:pkg": "node ../../scripts/build.mjs",
    "build:types": "tsc --project tsconfig.build.json --declaration --emitDeclarationOnly"
  },
  "dependencies": {}
}
