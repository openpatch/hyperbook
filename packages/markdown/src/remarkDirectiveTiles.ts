// Register directive nodes in mdast:
/// <reference types="mdast-util-directive" />
//
import { HyperbookContext } from "@hyperbook/types";
import { Comment, Element, Text } from "hast";
import { Root } from "mdast";
import { visit } from "unist-util-visit";
import { VFile } from "vfile";
import {
  expectContainerDirective,
  expectLeafDirective,
  isDirective,
  registerDirective,
} from "./remarkHelper";
import { Raw } from "mdast-util-to-hast";

export default (ctx: HyperbookContext) => () => {
  return (tree: Root, file: VFile) => {
    visit(tree, function (node) {
      if (isDirective(node)) {
        if (node.name !== "tiles") return;

        const data = node.data || (node.data = {});

        if (node.name === "tiles") {
          expectContainerDirective(node, file, "tiles");
          data.hName = "div";
          data.hProperties = {
            class: "directive-tiles",
          };
          registerDirective(file, "tiles", [], ["style.css"], ["tile"]);

          const tilesChildren: (Text | Comment | Element | Raw)[] = [];
          node.children.filter(isDirective).forEach((node) => {
            if (node.name !== "tile") {
              return;
            }
            const { title, size = "M", icon, href } = node.attributes || {};
            expectLeafDirective(node, file, "tile");

            if (!title) {
              file.fail(
                `[tiles] A title is required (${ctx.navigation.current?.href})`,
                node,
              );
            }

            const tileChildren: (Text | Comment | Element | Raw)[] = [
              {
                type: "element",
                tagName: "a",
                properties: {
                  class: "tile-title",
                  href: href ? ctx.makeUrl(href, "public") : undefined,
                },
                children: [
                  {
                    type: "text",
                    value: title,
                  },
                ],
              },
            ];

            if (icon) {
              tileChildren.push({
                type: "element",
                tagName: "img",
                properties: {
                  class: "tile-icon",
                  src: ctx.makeUrl(icon, "public"),
                },
                children: [],
              });
            }

            tilesChildren.push({
              type: "element",
              tagName: "li",
              properties: {
                class: `tile ${size} ${href ? "link" : ""}`,
              },
              children: tileChildren,
            });
          });

          data.hChildren = [
            {
              type: "element",
              tagName: "ul",
              properties: {
                class: "tiles",
              },
              children: tilesChildren,
            },
          ];
        }
      }
    });
  };
};
