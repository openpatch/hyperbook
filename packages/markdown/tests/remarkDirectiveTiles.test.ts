import { HyperbookContext } from "@hyperbook/types/dist";
import { describe, expect, it } from "vitest";
import rehypeStringify from "rehype-stringify";
import remarkParse from "remark-parse";
import remarkToRehype from "remark-rehype";
import { unified, PluggableList } from "unified";
import remarkDirective from "remark-directive";
import rehypeFormat from "rehype-format";
import remarkDirectiveRehype from "remark-directive-rehype";
import { ctx } from "./mock";
import remarkDirectiveTiles from "../src/remarkDirectiveTiles";

export const toHtml = (md: string, ctx: HyperbookContext) => {
  const remarkPlugins: PluggableList = [
    remarkDirective,
    remarkDirectiveRehype,
    remarkDirectiveTiles(ctx),
  ];

  return unified()
    .use(remarkParse)
    .use(remarkPlugins)
    .use(remarkToRehype)
    .use(rehypeFormat)
    .use(rehypeStringify, {
      allowDangerousCharacters: true,
      allowDangerousHtml: true,
    })
    .processSync(md);
};

describe("remarkDirectiveTiles", () => {
  it("should transform", async () => {
    expect(
      toHtml(
        `:::tiles

::tile{title="Hallo"}

::tile{title="A tile with a link" href="openpatch.org"}

::tile{title="A large tile" size="L"}

::tile{title="A small tile" size="S"}

::tile{title="A icon tile" icon="https://www.inf-schule.de/assets/img/icons/icon_algorithmen.png"}

:::
`,
        ctx,
      ).value,
    ).toMatchSnapshot();
  });
  it("should register directives", async () => {
    expect(
      toHtml(
        `:::tiles

::tile{title="Hallo"}

::tile{title="A tile with a link" href="openpatch.org"}

::tile{title="A large tile" size="L"}

::tile{title="A small tile" size="S"}

::tile{title="A icon tile" icon="https://www.inf-schule.de/assets/img/icons/icon_algorithmen.png"}

:::
`,
        ctx,
      ).data.directives?.["tiles"],
    ).toBeDefined();
  });
});
