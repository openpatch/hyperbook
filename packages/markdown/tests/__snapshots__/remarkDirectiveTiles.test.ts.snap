// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`remarkDirectiveTiles > should transform 1`] = `
"
<div class="directive-tiles">
  <ul class="tiles">
    <li class="tile M "><a class="tile-title">Hallo</a></li>
    <li class="tile M link"><a class="tile-title" href="/public/openpatch.org">A tile with a link</a></li>
    <li class="tile L "><a class="tile-title">A large tile</a></li>
    <li class="tile S "><a class="tile-title">A small tile</a></li>
    <li class="tile M "><a class="tile-title">A icon tile</a>
      <img class="tile-icon" src="/public/https:/www.inf-schule.de/assets/img/icons/icon_algorithmen.png">
    </li>
  </ul>
</div>
"
`;
